lib/WebService/Async/CustomerIO.pm

```diff
@@ -0,0 +1,313 @@
+
+Usage: C<< new(%params) -> obj >>
+
+Parameters:
+
+=over 4
+
+=item * C<site_id>
+
+=item * C<api_key>
+
+=back
+
+=cut
+
+sub configure {
+    my ($self, %args) = @_;
+    for my $k (qw(site_id api_key)) {
+        Carp::croak "Missing requeread argument: $k" unless exists $args{$k};
+        $self->{$k} = delete $args{$k} if exists $args{$k};
+    }
+
+    $self->{tracking_ratelimiter} ||= do {
+        my $rl =WebService::Async::CustomerIO::RateLimiter->new(
+            limit    => RPS_LIMIT_TRACKING,
+            interval => 1,
+        );
+        $self->add_child($rl);
+
+        $rl;
+    };
+    $self->{api_ratelimiter} ||= do {
+        my $rl = WebService::Async::CustomerIO::RateLimiter->new(
+            limit    => RPS_LIMIT_API,
+            interval => 1,
+        );
+        $self->add_child($rl);
+
+        $rl;
+    };
+
+    $self->next::method(%args);
+}
+
+=head2 site_id
+
+=cut
+
+sub site_id {shift->{site_id}}
+
+=head2 api_key
+
+=cut
+
+sub api_key {shift->{api_key}}
+
+=head2 API endpoints:
+
+There is 2 stable API for Customer.io, if you need to add a new method check 
+the L<documentation for API|https://customer.io/docs/api/> which endpoint 
+you need to use:
+
+=over 4
+
+=item * C<Tracking API> - Behavioral Tracking API is used to identify and track
+customer data with Customer.io.
+
+=item * C<Regular API> - Currently, this endpoint is only used for sending
+L<API triggered broadcasts|https://customer.io/docs/api-triggered-broadcast-setup>.
+
+=back
+
+=head2 tracking_request
+
+Sending request to Tracking API end point.
+
+Usage: C<< tracking_request($method, $uri, $data) -> future($data) >>
+
+=cut
+
+sub tracking_request {
+    my ($self, $method, $uri, $data) = @_;
+    return $self->tracking_ratelimiter->acquire->then(sub {
+        $self->_request($method, join(q{/} => (TRACKING_END_POINT, $uri)), $data);
+    });
+}
+
+=head2 api_request
+
+Sending request to Regular API end point.
+
+Usage: C<< api_request($method, $uri, $data) -> future($data) >>
+
+=cut
+
+sub api_request {
+    my ($self, $method, $uri, $data) = @_;
+
+    return $self->api_ratelimiter->acquire->then(sub {
+        $self->_request($method, join(q{/} => (API_END_POINT, $uri)), $data);
+    });
+}
+
+=head2 api_ratelimiter
+
+Getter returns RateLimmiter for regular API endpoint.
+
+=cut
+
+sub api_ratelimiter {shift->{api_ratelimiter}}
+
+=head2 tracking_ratelimiter
+
+Getter returns RateLimmiter for tracking API endpoint.
+
+=cut
+
+sub tracking_ratelimiter {shift->{tracking_ratelimiter}}
+
+sub _request {
+    my ($self, $method, $uri, $data) = @_;
+
+    my $body = $data             ? encode_json_utf8($data)
+             : $method eq 'POST' ? q{}
+             :                     undef;
+
+    return $self->_ua->do_request(
+            method  => $method,
+            uri     => $uri,
+            user    => $self->site_id,
+            pass    => $self->api_key,
+            !defined $body ? () : (
```

Comment:
why not other way around? `defined $body ? (...) : ()` more readable I guess

Reply:
this way, from my point of view is more convenient for indentation, therefore for reading as well.